feat: support toolchain export

Halimao opened the pull request:
close #322 

Support `fuelup toolchain export`

![image](https://github.com/FuelLabs/fuelup/assets/25278203/0d2809c5-02ea-403b-bee0-0a3ae1bb27ea)

![image](https://github.com/FuelLabs/fuelup/assets/25278203/59431ea5-e300-42ce-86b9-4c6ea977ed5c)

![image](https://github.com/FuelLabs/fuelup/assets/25278203/67e872b9-e7c2-45c2-98dc-676494e80872)

![image](https://github.com/FuelLabs/fuelup/assets/25278203/1ebbf74a-98f7-48a4-8f2f-65e0f8cfc444)

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,98 @@
+    let toolchain_override = ToolchainOverride {
+        cfg: OverrideCfg::new(
+            ToolchainCfg {
+                channel: toolchain_override::Channel::from_str(&toolchain_name).unwrap(),
```
I don't think this will work for custom toolchains. The channel must be one of the distributed channels; it can't be an arbitrary name of a custom toolchain. See `deserialize_channel`

Some options are: 
1. We could default the channel to latest. But, that's making an assumption.
2. We can ask the user to provide a valid channel and then continue the export.
3. We can restrict users to only be able to export their toolchain if they're using a distributed toolchain. This would still be useful for just generating a fuel-toolchain.toml file, as `forc init` currently does not generate this file.
4. In the case where a user supplies `--force` and the file already exists, we could use the `channel` from the previous file, and only override the components where they differ from the user's default toolchain. 

I don't have a strong opinion on this, but I'd lean towards doing 2 & 4.

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,98 @@
+    let ExportCommand { name, force} = command;
+    let path = PathBuf::from("./").join(FUEL_TOOLCHAIN_TOML_FILE);
+    if !force && path.exists() {
+        bail!("{FUEL_TOOLCHAIN_TOML_FILE} already exists");
```
Let's do a similar behavior to what's being done in https://github.com/FuelLabs/forc-wallet/pull/148

That is, if `--force` is not supplied, ask the user if they wish to override it. If it is, give the user a warning that the file is being overwritten.

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,98 @@
+
+pub fn export(command: ExportCommand) -> Result<()> {
+    let ExportCommand { name, force} = command;
+    let path = PathBuf::from("./").join(FUEL_TOOLCHAIN_TOML_FILE);
```
We should be smarter about locating the toml file. The user should be able to run this command from any level within the forc project.

You can use `get_fuel_toolchain_toml` for this.

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,98 @@
+    Ok(())
+}
+
+fn get_exec_version(component_executable: &Path) -> Result<Version> {
```
How is this different from `exec_version`? Let's move that function to a file called `util/version.rs`, make it public, and import it to use it here.

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    bail!("could not show version: {}", component_executable.display());
+}
```
Where are the tests?

sdankel reviewed (changes requested):
Thanks for your PR! 🚀

sdankel commented on src/commands/toolchain.rs:
```diff
@@ -42,6 +46,16 @@ pub struct UninstallCommand {
+    /// Toolchain to export, [possible values: latest, beta-1, beta-2, beta-3, beta-4, nightly].
+    /// Your currently active toolchain will be exported if name isn't specified
+    pub name: Option<String>,
+    /// Set true to disregard the existing TOML output file
```
```suggestion
    /// Forces exporting the toolchain, replacing any existing toolchain override file
```

sdankel commented on src/commands/toolchain.rs:
```diff
@@ -42,6 +46,16 @@ pub struct UninstallCommand {
+#[derive(Debug, Parser)]
+pub struct ExportCommand {
+    /// Toolchain to export, [possible values: latest, beta-1, beta-2, beta-3, beta-4, nightly].
+    /// Your currently active toolchain will be exported if name isn't specified
```
`Toolchain::from_settings()` gives the default toolchain, not the currently active one.

```suggestion
    /// The default toolchain will be exported if name isn't specified
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,267 @@
+        input_toolchain_name = String::from(input_toolchain_name.trim());
+        if toolchain_override::Channel::from_str(&input_toolchain_name).is_err() {
+            bail!(
+                "Invalid channel '{}', expected one of <latest-YYYY-MM-DD|nightly-YYYY-MM-DD|beta-1|beta-2|beta-3|beta-4>.",
```
In the interest of DRY, let's move this to a constant

```
const INVALID_CHANNEL_STR: &str = "Invalid channel '{}', expected one of <latest-YYYY-MM-DD|nightly-YYYY-MM-DD|beta-1|beta-2|beta-3|beta-4>.";
```

Halimao replied:
Done

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,267 @@
+        input_toolchain_name = String::from(input_toolchain_name.trim());
+        if toolchain_override::Channel::from_str(&input_toolchain_name).is_err() {
+            bail!(
+                "Invalid channel '{}', expected one of <latest-YYYY-MM-DD|nightly-YYYY-MM-DD|beta-1|beta-2|beta-3|beta-4>.",
```
This makes me think we should also accept an optional `--channel` argument so the user doesn't have to do this interactively.

Halimao replied:
@sdankel Done

Halimao replied:
Note: `toolchain_override::Channel::from_str` logic about validating distributed channels updated, as some invalid channels like "my-custom", "latest-blabla" maybe passed by mistake

![image](https://github.com/FuelLabs/fuelup/assets/25278203/43c49bc1-e2ef-4094-a84a-caf80f71fa06)

sdankel reviewed (commented):
Thanks for adding the tests! I left a few more comments, and will test this out next week.

sdankel commented on src/toolchain_override.rs:
```diff
@@ -69,7 +69,7 @@ where
             Err(Error::invalid_value(
                 serde::de::Unexpected::Str(&channel_str),
-                &"one of <latest-YYYY-MM-DD|nightly-YYYY-MM-DD|beta-1|beta-2|beta-3|beta-4>",
+                &format!("one of {}", VALID_CHANNEL_STR).as_str(),
```
👍

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+            }
+        }
+    }
+    let mut export_channel = toolchain_name.clone();
+    if let Some(channel) = channel {
+        export_channel = channel;
+    }
```
```suggestion
    let export_channel = channel.unwrap_or(toolchain_name.clone());
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+            }
+        }
+    }
+    let mut toolchain_name = Toolchain::from_settings()?.name;
+    if let Some(name) = name {
+        toolchain_name = name;
+    }
```
```suggestion
    let toolchain_name = name.unwrap_or(Toolchain::from_settings()?.name);
```

sdankel commented on src/commands/toolchain.rs:
```diff
@@ -42,6 +50,20 @@ pub struct UninstallCommand {
+    /// Toolchain to export, [possible values: latest, beta-1, beta-2, beta-3, beta-4, nightly].
+    /// The default toolchain will be exported if name isn't specified
+    pub name: Option<String>,
+    /// Channel to export, [possible values: latest-YYYY-MM-DD, nightly-YYYY-MM-DD, beta-1, beta-2, beta-3, beta-4].
+    #[clap(value_parser = channel_allowed)]
```
clap will add the possible values to the help message for you:

```suggestion
    /// Channel to export, [possible values: latest-YYYY-MM-DD, nightly-YYYY-MM-DD, beta-1, beta-2, beta-3, beta-4].
    #[clap(value_parser = clap::builder::PossibleValuesParser::new(["latest-YYYY-MM-DD", "nightly-YYYY-MM-DD", "beta-1", "beta-2", "beta-3", "beta-4"]))]
```

Halimao replied:
@sdankel  Hi sir, I think using `clap::builder::PossibleValuesParser` would cause unexpected behaviour. Will clap parse "YYYY-MM-DD" correctly? I tried to provide `latest-2023-12-13` but failed.

![image](https://github.com/FuelLabs/fuelup/assets/25278203/c9e24d04-1a83-4209-b8b8-8175c307514a)

sdankel commented on src/commands/toolchain.rs:
```diff
@@ -64,12 +86,27 @@ fn name_allowed(s: &str) -> Result<String> {
     }
 }
 
+fn channel_allowed(s: &str) -> Result<String> {
+    if s.is_empty() {
+        return Ok(s.to_string());
+    }
+    if toolchain_override::Channel::from_str(s).is_err() {
+        bail!(
+            "Invalid channel '{}', expected one of {}.",
+            s,
+            VALID_CHANNEL_STR,
+        );
+    }
+    Ok(s.to_string())
+}
+
```
```suggestion
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+        toolchain_info_path = path;
+        if force {
+            println_warning(&format!(
+                "Because the `--force` argument was supplied, the toolchain info file at {} will be removed.",
```
```suggestion
                "Because the `--force` argument was supplied, this file will be overwritten: {}",
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+            fs::remove_file(&toolchain_info_path).unwrap();
+        } else {
+            println_warning(&format!(
+                "There is an existing toolchain info file at {}. \
```
```suggestion
                "There is an existing toolchain override file at {}. \
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+                fs::remove_file(&toolchain_info_path).unwrap();
+            } else {
+                bail!(
+                    "Failed to create a new toolchain info file at {} \
+                    because a toolchain info file already exists at that location.",
```
```suggestion
                    "Failed to export toolchain \
                    because a toolchain override file already exists at {}.",
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+                    if !plugin.is_main_executable() {
+                        for executable in plugin.executables.iter() {
+                            let plugin_executable = export_toolchain.bin_path.join(executable);
+                            if let Ok(version) = exec_version(plugin_executable.as_path()) {
```
Is `as_path()` needed here?

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+
+    let mut version_map: HashMap<String, Version> = HashMap::new();
+    for component in Components::collect_exclude_plugins()? {
+        let component_executable = export_toolchain.bin_path.join(&component.name);
+        if let Ok(version) = exec_version(component_executable.as_path()) {
+            version_map.insert(component.name.clone(), version);
+        };
```
```suggestion
        exec_version(export_toolchain.bin_path.join(&component.name))
            .map(|version| version_map.insert(component.name.clone(), version));
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,320 @@
+            let forc_executables = component.executables;
+            for plugin in Components::collect_plugins()? {
+                if !forc_executables.contains(&plugin.name) {
+                    if !plugin.is_main_executable() {
+                        for executable in plugin.executables.iter() {
+                            let plugin_executable = export_toolchain.bin_path.join(executable);
+                            if let Ok(version) = exec_version(plugin_executable.as_path()) {
+                                version_map.insert(executable.clone(), version);
+                            };
+                        }
+                    } else {
+                        let plugin_executable = export_toolchain.bin_path.join(&plugin.name);
+                        if let Ok(version) = exec_version(plugin_executable.as_path()) {
+                            version_map.insert(plugin.name.clone(), version);
+                        };
+                    }
```
I don't think we don't need to check plugin.is_main_executable() because the behavior is the same.

```suggestion
                    plugin.executables.iter().for_each(|executable| {
                        exec_version(export_toolchain.bin_path.join(executable))
                            .map(|version| version_map.insert(executable.clone(), version));
                    });
```

sdankel reviewed (commented):
Looking good, left some nits.

Halimao commented:
@sdankel Hi sir, looking forward for your review again😀

JoshuaBatty commented on src/ops/fuelup_toolchain/export.rs:
Thanks @Halimao. Would you mind going back over and removing the unwraps and either handle the error cases, or bail with a message to user?

Halimao replied:
> Thanks @Halimao. Would you mind going back over and removing the unwraps and either handle the error cases, or bail with a message to user?

Hi @JoshuaBatty , addressed, thanks for your review~

JoshuaBatty replied:
Thanks, i'll take another look

Braqzen commented on src/commands/toolchain.rs:
```diff
@@ -42,6 +50,20 @@ pub struct UninstallCommand {
 
+#[derive(Debug, Deserialize, Parser)]
+pub struct ExportCommand {
+    /// Toolchain to export, [possible values: latest, beta-1, beta-2, beta-3, beta-4, nightly].
+    /// The default toolchain will be exported if name isn't specified
+    pub name: Option<String>,
```
Why does this not allow custom toolchains?

Halimao replied:
> Why does this not allow custom toolchains?

Nop, there is no limit for name to be non-custom toolchains, but the channel does

Braqzen commented on src/commands/toolchain.rs:
```diff
@@ -42,6 +50,20 @@ pub struct UninstallCommand {
+    /// Toolchain to export, [possible values: latest, beta-1, beta-2, beta-3, beta-4, nightly].
+    /// The default toolchain will be exported if name isn't specified
+    pub name: Option<String>,
+    /// Channel to export, [possible values: latest-YYYY-MM-DD, nightly-YYYY-MM-DD, beta-1, beta-2, beta-3, beta-4].
```
Does this also have a default if not specified?

Halimao replied:
Just as described here "The default toolchain will be exported if name isn't specified" , the default toolchain by `fuelup default` would be used for.

![image](https://github.com/FuelLabs/fuelup/assets/25278203/7f7c2944-0b37-4f9b-ba11-5a41a01984df)

Braqzen commented on src/commands/toolchain.rs:
```diff
@@ -64,12 +86,27 @@ fn name_allowed(s: &str) -> Result<String> {
+            VALID_CHANNEL_STR,
+        );
+    }
+    Ok(s.to_string())
```
Could we return a string slice too?

Halimao replied:
Why should this be returned a string slice, do you means that we need to support for export with multi channels?

Braqzen replied:
Little less work for the compiler if we don't turn it into a String. No biggie

Braqzen commented on src/commands/toolchain.rs:
```diff
@@ -64,12 +86,27 @@ fn name_allowed(s: &str) -> Result<String> {
     }
 }
 
+fn channel_allowed(s: &str) -> Result<String> {
```
What is `s`?

Also a bit confused about the returning of an empty `String` version of `s`

Halimao replied:
`channel_allowed` is used as clap `value_parser` for channel, so s means the `channel` option value. And as `--channel` is optional, we should return empty `String`, for unwrap `channel: Option<String>` in `Export`

Braqzen replied:
Would it be more apt to rename to `channel`?

Halimao replied:
I will update it later👌

Halimao replied:
@Braqzen Done

Braqzen commented on src/commands/toolchain.rs:
```diff
@@ -64,12 +86,27 @@ fn name_allowed(s: &str) -> Result<String> {
         ToolchainCommand::New(command) => new(command)?,
         ToolchainCommand::Uninstall(command) => uninstall(command)?,
         ToolchainCommand::ListRevisions(command) => list_revisions(command)?,
+        ToolchainCommand::Export(command) => export(command, stdin().lock())?,
```
What's the reason for locking the `stdin` here?

Halimao replied:
I have updated to `BufReader::new(stdin())`

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+                &toolchain_info_path.display(),
+            ));
+            let mut need_replace = String::new();
+            if reader.read_line(&mut need_replace).is_err() {
```
I see. This is probably why we locked above

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+                    bail!("failed to remove file {}", &toolchain_info_path.display());
+                };
+            } else {
+                bail!(
+                    "Failed to export toolchain \
+                    because a toolchain override file already exists at {}.",
+                    &toolchain_info_path.display(),
+                );
```
I think a more precise message would be to say that we couldn't remove it because the user has declined the replacement rather than it existed

Halimao replied:
@sdankel What do you think of this, before [suggestion](https://github.com/FuelLabs/fuelup/pull/526#discussion_r1424644440)

sdankel replied:
@Braqzen do you have a suggestion for what the error message should be? It's hard to say that without being too wordy. Maybe something like "Canceled toolchain export." would be enough.

Braqzen replied:
> Cancelled toolchain export

Is good. It simply states that the action has been discontinued so once the prompt pops up for the user they know that there has been no error. The operation has ended.

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+        }
+    }
+    // mock setting.toml
+    fn create_toolchain_settings_file() {
```
I believe these helpers touch the disk i.e. create and write a file and then remove it.
Does Rust have a library for doing this all in-memory instead? If it does then we could pass the handle around and inspect it instead of reading / writing when testing and potentially leaving behind some files if something exits early.

This is coming from the perspective of having done something like this in Python before

Halimao replied:
I'm not sure. But `Export`function inner will read file from real disk not the memory.

sdankel replied:
I think it's okay for the tests to read/write to disk, but the tests should clean up after themselves.

Braqzen reviewed (commented):
Question about the exporting format in the images. It looks to be Json-ish.

[Here](https://github.com/FuelLabs/sway-applications/blob/354bac23f11db6b2f89df53b46a8afe43c31911d/AMM/project/fuel-toolchain.toml#L1-L6) is an example of a toolchain file and they look to be formatted differently.

Am I overlooking something or are the formats different when exported?

Halimao commented:
> Question about the exporting format in the images. It looks to be Json-ish.
> 
> [Here](https://github.com/FuelLabs/sway-applications/blob/354bac23f11db6b2f89df53b46a8afe43c31911d/AMM/project/fuel-toolchain.toml#L1-L6) is an example of a toolchain file and they look to be formatted differently.
> 
> Am I overlooking something or are the formats different when exported?

Exporting format is standard toml format as well, you can check it here: https://www.toml-lint.com/ .
![image](https://github.com/FuelLabs/fuelup/assets/25278203/1c3d68f2-cc30-407d-98e7-4b288083d7cd)

Halimao commented on src/util/version.rs:
```diff
@@ -0,0 +1,30 @@
+            if component_executable.exists() {
+                bail!("execution error - {}", e);
+            } else {
+                bail!("not found");
```
See screenshot below, just like "Error getting version string", not found will show after component name, meas this component not found. I think it's fine.

![image](https://github.com/FuelLabs/fuelup/assets/25278203/3b87bc00-5dfa-46c2-bcec-ff53200f6246)

Halimao commented:
@Braqzen @sdankel Hi sir, is there anything that needs to be added for this pr?

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+            ));
+            let mut need_replace = String::new();
+            if reader.read_line(&mut need_replace).is_err() {
+                bail!("failed to read user input");
```
```suggestion
                bail!("Failed to read user input");
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+            }
+            if need_replace.trim() == "y" {
+                if fs::remove_file(&toolchain_info_path).is_err() {
+                    bail!("failed to remove file {}", &toolchain_info_path.display());
```
```suggestion
                    bail!("Failed to remove file {}", &toolchain_info_path.display());
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+                    bail!("failed to remove file {}", &toolchain_info_path.display());
+                };
+            } else {
+                bail!(
+                    "Failed to export toolchain \
+                    because a toolchain override file already exists at {}.",
+                    &toolchain_info_path.display(),
+                );
```
```suggestion
                println_warning("Canceled toolchain export");
                return Ok(());
```

@Braqzen thoughts?

Braqzen replied:
Probably okay.

I am wondering about general tool design here. Specifically, does it make sense to print a "warning" (maybe it would be better to rename to message) and then to exit with an `Ok(())` if we choose to exit the program early.

In my mind it makes more sense to propagate a specific error all the way up and something would catch that and then handle it by ignoring it or something.

Returning with an `ok` before the intended completion of the workflow is strange to me but the user did choose to cancel so perhaps that can be considered as `ok`.

It's probably too much work to implement this change throughout so it's probably okay as it is although I would generalize to `println_message` unless we have different `println_method`s already

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+        ));
+        let mut input_channel_name = String::new();
+        if reader.read_line(&mut input_channel_name).is_err() {
+            bail!("failed to read user input");
```
nit
```suggestion
            bail!("Failed to read user input");
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+    };
+    let document = toolchain_override.to_toml();
+    if std::fs::write(toolchain_override.path, document.to_string()).is_err() {
+        bail!("failed to write {}", toolchain_info_path.display());
```
```suggestion
        bail!("Failed to write {}", toolchain_info_path.display());
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+        {
+            Ok(())
+        } else {
+            bail!("unexpected channel");
```
```suggestion
            bail!("Unknown channel \"{channel_name}\"");
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+                &toolchain_info_path.display(),
+            ));
+            if fs::remove_file(&toolchain_info_path).is_err() {
+                bail!("failed to remove file {}", &toolchain_info_path.display());
```
```suggestion
                bail!("Failed to remove file {}", &toolchain_info_path.display());
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+    }
+
+    info!(
+        "exported '{}' into '{}'",
```
```suggestion
        "Exported toolchain \"{}\" to file: /"{}/"",
```

sdankel commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,331 @@
+            file.write_all(b"default_toolchain = \"latest\"").unwrap();
+        }
+    }
+    fn check_toolchain_info_with_channel(channel_name: &String) -> Result<()> {
+        let toolchain_override = ToolchainOverride::from_project_root().unwrap();
+        if toolchain_override
+            .cfg
+            .toolchain
+            .channel
+            .to_string()
+            .eq(channel_name)
+        {
+            Ok(())
+        } else {
+            bail!("unexpected channel");
+        }
+    }
```
This is a good use case for [matches](https://doc.rust-lang.org/std/macro.matches.html)

```suggestion
    fn assert_channel_name(expected: &String) {
        let toolchain_override = ToolchainOverride::from_project_root().expect("toolchain override");
        let actual = toolchain_override
            .cfg
            .toolchain
            .channel
            .to_string();
        assert!(matches!(expected, actual));
    }
```

Braqzen reviewed (commented):
Running `toolchain export` without any args / options I see the following

![1](https://github.com/FuelLabs/fuelup/assets/103777923/8c48814f-257c-4ab1-bbed-30c424c873a3)

I would move the message asking for input onto a new line instead of having a single long line.
The default is not automatically exported, it prompts for input.

The `invalid channel` message comes from a break in backwards compatibility. I haven't followed the changes to know how we manage new toolchains etc. so I'm unsure if this `x86` channel is valid but I have updated fuelup and installed `beta-4` so I assume that it's still valid.
All that to say, we are forcing a new format, have we considered the current / old format and whether we ought to support that?

Running `toolchain export` with `beta-4` I get a file containing the toolchain and components however when I run `latest-YYYY-MM-DD` I get the following error

![1](https://github.com/FuelLabs/fuelup/assets/103777923/c7275d73-1f7b-4fbe-aa0d-1c04dcf47dee)

We ought to handle the difference between a valid channel that the user can type in ex. the beta-X vs the example formats which include a date (as seen above).

According to the possible argument values I should be able to use `latest` however when I do that it tells me that it's an invalid channel.

![1](https://github.com/FuelLabs/fuelup/assets/103777923/7fcab4d1-1423-439b-9456-3a8e6068d842)

Overriding an existing toolchain uses the incorrect toolchain. It ignores my input and uses the default now? When I check the content of the file it has `beta-4` so it only prints the incorrect toolchain.

![1](https://github.com/FuelLabs/fuelup/assets/103777923/cc041ed0-e1ea-4ead-a56f-548da14172ea)

Halimao commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,319 @@
+            if reader.read_line(&mut need_replace).is_err() {
+                bail!("Failed to read user input");
+            }
+            if need_replace.trim() == "y" {
```
> I've recently done a little `stdin` reading. Does this remove spaces and covers new lines? \n and \r?

I think this could be confirmed in the unit test

![image](https://github.com/FuelLabs/fuelup/assets/25278203/c6a20acd-b65a-4189-aabd-c40037447caf)

Halimao commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,319 @@
+    let mut toolchain_info_path = PathBuf::from("./").join(FUEL_TOOLCHAIN_TOML_FILE);
+    if let Some(path) = get_fuel_toolchain_toml() {
+        toolchain_info_path = path;
+        if force {
```
Done

Halimao commented on src/commands/toolchain.rs:
```diff
@@ -1,12 +1,18 @@
-use anyhow::{bail, Result};
+use anyhow::{bail, Ok, Result};
```
Removed

Halimao commented:
> Running `toolchain export` without any args / options I see the following
> 
> ![1](https://private-user-images.githubusercontent.com/103777923/295647998-8c48814f-257c-4ab1-bbed-30c424c873a3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5Mzg2MjIsIm5iZiI6MTcwNDkzODMyMiwicGF0aCI6Ii8xMDM3Nzc5MjMvMjk1NjQ3OTk4LThjNDg4MTRmLTI1N2MtNGFiMS1iYmVkLTMwYzQyNGM4NzNhMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMVQwMTU4NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYzBhZTYzZjViOTg2MjkxMWUzYmJmNTA1YjcxMzI3ODg4ZmQyMmYzMWQzMjQ5YWVlYzZkZWVjMDZmNmQyMTJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.oZIN5UiEQBHrZ1lZ5zk_6dWyzKX9snvJ-3dZhB7c4Do)
> 
> I would move the message asking for input onto a new line instead of having a single long line. The default is not automatically exported, it prompts for input.
> 
> The `invalid channel` message comes from a break in backwards compatibility. I haven't followed the changes to know how we manage new toolchains etc. so I'm unsure if this `x86` channel is valid but I have updated fuelup and installed `beta-4` so I assume that it's still valid. All that to say, we are forcing a new format, have we considered the current / old format and whether we ought to support that?
> 
> Running `toolchain export` with `beta-4` I get a file containing the toolchain and components however when I run `latest-YYYY-MM-DD` I get the following error
> 
> ![1](https://private-user-images.githubusercontent.com/103777923/295654173-c7275d73-1f7b-4fbe-aa0d-1c04dcf47dee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5Mzg2MjIsIm5iZiI6MTcwNDkzODMyMiwicGF0aCI6Ii8xMDM3Nzc5MjMvMjk1NjU0MTczLWM3Mjc1ZDczLTFmN2ItNGZiZS1hYTBkLTFjMDRkY2Y0N2RlZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMVQwMTU4NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YTVlNjY2MWVkNGMzNDIxMTNiOWE1OWNiYmI2ZjI2NWI2ZDc1MTk2NzNlYjEzYTgxYjQwNWEwNzU0OTA1NmEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.BxEcxlNc2Ap0Qn6GgWl105Ye4Mzr09ytjHCpnZpB2Dw)
> 
> We ought to handle the difference between a valid channel that the user can type in ex. the beta-X vs the example formats which include a date (as seen above).
> 
> According to the possible argument values I should be able to use `latest` however when I do that it tells me that it's an invalid channel.
> 
> ![1](https://private-user-images.githubusercontent.com/103777923/295654669-7fcab4d1-1423-439b-9456-3a8e6068d842.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5Mzg2MjIsIm5iZiI6MTcwNDkzODMyMiwicGF0aCI6Ii8xMDM3Nzc5MjMvMjk1NjU0NjY5LTdmY2FiNGQxLTE0MjMtNDM5Yi05NDU2LTNhOGU2MDY4ZDg0Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMVQwMTU4NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YjUzZjE0NDc5YTBkNDU0OTdiYzc0ODdjMzZiZTMzYWM1NTY2NGM3MDJmYmE4MGYxYTA3MmFlNzkwM2I4OGY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CUVjysEIbOSoWvtiCV8oD24iBOY0LbhnwUZW_ECfDrU)
> 
> Overriding an existing toolchain uses the incorrect toolchain. It ignores my input and uses the default now? When I check the content of the file it has `beta-4` so it only prints the incorrect toolchain.
> 
> ![1](https://private-user-images.githubusercontent.com/103777923/295655445-cc041ed0-e1ea-4ead-a56f-548da14172ea.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5Mzg2MjIsIm5iZiI6MTcwNDkzODMyMiwicGF0aCI6Ii8xMDM3Nzc5MjMvMjk1NjU1NDQ1LWNjMDQxZWQwLWUxZWEtNGVhZC1hNTZmLTU0OGRhMTQxNzJlYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDExMVQwMTU4NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNzNlMjBhNjEwNDQ2ZmFiZWY5MDFjYzYxNGJiZGVmMDlmZGZiZTY0NDVlMGIzZDI0MmU1NmJmOTM5ZDNlNzk5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.2FxoG8TibyRoETVNiKw0cazWa_keGzHkLWgJ_WtXnwg)

I think this should be discussed with @sdankel first. According to the confusion listed above, it seems that you have mixed up the `toolchain name` and `toolchain channel`.

Such as

> Running `toolchain export` without any `args / options` shows the promotions that ask for a valid channel.

Reference to the [comments by @sdankel ](https://github.com/FuelLabs/fuelup/pull/526#discussion_r1412376018)
This is caused by the default toolchain name you have(latest-x86_64_unknown...) is not a valid channel but a valid toolchain name. When there is no --channel provided, we will use the default toolchain name as the channel, and then check the validation of `channel`

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+                        bail!("Failed to remove file {}", &toolchain_info_path.display());
+                    };
+                } else {
+                    println_warning("Canceled toolchain export");
```
```suggestion
                    println_warning("Cancelled toolchain export");
```

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+    fn remove_toolchain_info() {
+        if let Some(fuel_toolchain_toml_file) = get_fuel_toolchain_toml() {
+            if fuel_toolchain_toml_file.exists() {
+                fs::remove_file(fuel_toolchain_toml_file).unwrap();
```
Do we want to change the `unwrap()`s to `expect(<msg>)` for clearer failures in tests or are we fine with unwrapping and crashing?

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+    fn assert_channel_name(expected: &String) {
+        let toolchain_override =
+            ToolchainOverride::from_project_root().expect("toolchain override");
+        let _actual = toolchain_override.cfg.toolchain.channel.to_string();
+        assert!(matches!(expected, _actual));
```
```suggestion
        let actual = toolchain_override.cfg.toolchain.channel.to_string();
        assert!(matches!(expected, actual));
```

Halimao replied:
reported warning by rustc lint 
> if this is intentional, prefix it with an underscore: `_actual`rustc(unused_variables)

sdankel replied:
That's odd, maybe the compiler doesn't recognize the usage inside the `matches` macro. It would be better to use `assert_eq` here anyway.

Halimao replied:
> That's odd, maybe the compiler doesn't recognize the usage inside the `matches` macro. It would be better to use `assert_eq` here anyway.

@sdankel Hi sir, suggestion changes addressed!

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+    }
+
+    #[test]
+    #[serial]
```
What does this annotation do?

Halimao replied:
> What does this annotation do?

For tests that could be guaranteed to be executed in serial, as there are `file creation and deletion`, executing in parallel is non-guaranteed for `file creation and deletion` and causes failed

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+
+    #[test]
+    #[serial]
+    fn export_toolchain_with_exists_toolchain_info_with_cancel() {
```
```suggestion
    fn export_toolchain_with_existing_toolchain_info_then_cancel() {
```

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+    #[test]
+    #[serial]
+    #[should_panic]
+    fn export_toolchain_with_invalid_channel_provided_throws_err() {
```
```suggestion
    fn export_toolchain_with_invalid_channel() {
```

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+    #[test]
+    #[serial]
+    #[should_panic]
+    fn export_toolchain_with_invalid_channel_inputted_throws_err() {
```
```suggestion
    fn export_toolchain_without_channel() {
```

Braqzen commented on src/ops/fuelup_toolchain/export.rs:
```diff
@@ -0,0 +1,322 @@
+
+    #[test]
+    #[serial]
+    fn export_toolchain_with_exists_toolchain_info() {
```
```suggestion
    fn export_toolchain_with_forced_overwrite() {
```

Braqzen replied:
Looking at this and then the following test below I would split them into unit tests because there's more than 1 test case in each.

Braqzen reviewed (commented):
I would also split the tests a bit further to group them in order to reduce the cognitive load and make it easier to read.

I would opt for either putting all the success cases at the top and panics at the bottom or consider putting the success cases in a `mod success` and the panics in a `mod panic`.

After the tests are split I'll look over test coverage because I haven't exactly checked if all cases are covered

Halimao commented:
> I would also split the tests a bit further to group them in order to reduce the cognitive load and make it easier to read.
> 
> I would opt for either putting all the success cases at the top and panics at the bottom or consider putting the success cases in a `mod success` and the panics in a `mod panic`.
> 
> After the tests are split I'll look over test coverage because I haven't exactly checked if all cases are covered

Addressed!

sdankel commented:
This produces a file that looks like this:

```
toolchain = { channel = "beta-3" }
components = { forc = "0.46.1", forc-explore = "0.28.1", forc-tx = "0.46.1", forc-deploy = "0.46.1", forc-wallet = "0.3.0", fuel-core = "0.20.5", forc-run = "0.46.1" }
```

But we want it formatted like this:

```
[toolchain]
channel = "beta-3"
    
[components]
forc-wallet = "0.3.0",
forc-deploy = "0.46.1",
forc-run = "0.46.1",
forc-explore = "0.28.1",
forc-tx = "0.46.1",
forc = "0.46.1",
fuel-core = "0.20.5"
```

sdankel commented:
I fixed the formatting issue and added tests for it. However I noticed it still pulls the wrong name for the channel by default. When I'm on beta-3, it pulls the full toolchain name rather than the shortened dist name for the channel.

This test should pass:

```
    #[test]
    #[serial]
    fn test_export_beta3_full() {
        create_settings_file();
        create_toolchain_info();

        export(
            ExportCommand {
                name: None,
                channel: Some("beta-3-x86_64-apple-darwin".to_string()),
                force: true,
            },
            &INPUT_NO[..],
        )
        .expect("should succeed");
        assert_toolchain_info(BETA_3_FILE_CONTENTS);
    }
```

Halimao commented:
> This test should pass:
> 
> ```
>             ExportCommand {
>                 name: None,
>                 channel: Some("beta-3-x86_64-apple-darwin".to_string()),
>                 force: true,
>             },
>             &INPUT_NO[..],
>         )
> ```

One of the valid channel should be `<latest-YYYY-MM-DD|nightly-YYYY-MM-DD|beta-1|beta-2|beta-3|beta-4>`, not sure if this("beta-3-x86_64-apple-darwin" is not a valid channel) is the reason why this test shouldn't pass?

Halimao commented:
> What I'm saying is if I'm on beta-3, the toolchain name is beta-3-x86_64-apple-darwin. If I just run fuelup toolchain export, I expect it to work. Currently it gives me the error that beta-3-x86_64-apple-darwin is not a valid channel. I think this is because of https://github.com/FuelLabs/fuelup/issues/545

I'm sorry, but this is expected IMO. As we discussed before, 
1. if no toolchain name is provided(run only `fuelup toolchain export`), we will use the default toolchain name, in your local env, that's `beta-3-x86_64-apple-darwin`.
2. Also, as there is no `--channel` provided, we will try to use the toolchain name as channel first. 
3. Then check the channel name, it will prompt user "beta-3-x86_64-apple-darwin is not a valid channel, please input a valid one"
4. If user input `beta-3`, then this will success as `beta-3` is a valid channel, otherwise, abort the process as input channel is invalid

sdankel commented:
> if no toolchain name is provided(run only fuelup toolchain export), we will use the default toolchain name, in your local env, that's beta-3-x86_64-apple-darwin.

Yes, and we should parse out `beta-3` from the toolchain name using `DistToolchainDescription::from_str` once https://github.com/FuelLabs/fuelup/pull/566 lands.

JoshuaBatty commented:
This has been implemented in #777
